chaos_genius/views/anomaly_data_view.py

```diff
@@ -425,6 +357,53 @@ def kpi_anomaly_retraining(kpi_id):
         return jsonify({"msg": f"retraining failed for KPI: {kpi_id}, KPI id is None"})
 
 
+def _get_dimensions_values(
+    kpi_id: int, end_date: datetime, period=90
+) -> Dict[str, List[str]]:
+    """Creates a dictionary of KPI dimension and their values.
+
+    :param kpi_id: ID of the KPI
+    :type kpi_id: int
+    :return dimension_values_dict: dictionary of {dimension:list(vals)}
+    :rtype: dict
+    """
+    start_date = pd.to_datetime(end_date) - timedelta(days=period)
+    start_date = start_date.strftime("%Y-%m-%d %H:%M:%S")
+    end_date = end_date.strftime("%Y-%m-%d %H:%M:%S")
+
+    # Get unique list of subdims and values from DB
+    results = (
+        db.session.query(
+            func.distinct(AnomalyDataOutput.series_type)
+        )
+        .filter(
+            (AnomalyDataOutput.kpi_id == kpi_id)
+            & (AnomalyDataOutput.data_datetime >= start_date)
+            & (AnomalyDataOutput.data_datetime <= end_date)
+            & (AnomalyDataOutput.anomaly_type == "subdim")
+        ).all()
+    )
```

Comment:
There's a type error on line 372: `Expression of type "str" cannot be assigned to declared type "datetime" `. It's because `end_date` was declared with a specific type (`datetime`) in the function parameters. Looks like the type checker does not allow rebinding an identifier with a different type. The only solution I found was using a different name for the variable (and I changed the name of `start_date_str` to be consistent):

```suggestion
    start_date_str = start_date.strftime("%Y-%m-%d %H:%M:%S")
    end_date_str = end_date.strftime("%Y-%m-%d %H:%M:%S")

    # Get unique list of subdims and values from DB
    results = (
        db.session.query(
            func.distinct(AnomalyDataOutput.series_type)
        )
        .filter(
            (AnomalyDataOutput.kpi_id == kpi_id)
            & (AnomalyDataOutput.data_datetime >= start_date_str)
            & (AnomalyDataOutput.data_datetime <= end_date_str)
            & (AnomalyDataOutput.anomaly_type == "subdim")
        ).all()
    )
```

Reply:
Wait. Does it really need to be a string? I think SQLalchemy can work with a `datetime` too.

Reply:
Not sure if it needs to be a string, but I think there were some issues with the datetime formats and the queries giving incomplete data at boundaries. So I think Kartikay just made it uniform across all functions.